chef/cookbooks/nova/recipes/ceph.rb

```diff
@@ -0,0 +1,185 @@
+return unless cinder_controller.length > 0
+
+# First loop to find if we have internal/external cluster
+cinder_controller[0][:cinder][:volumes].each_with_index do |volume, volid|
+  next unless volume[:backend_driver] == "rbd"
+
+  has_internal ||= true if volume[:rbd][:use_crowbar]
+  has_external ||= true unless volume[:rbd][:use_crowbar]
+end
+
+if has_internal
+  ceph_env_filter = " AND ceph_config_environment:ceph-config-default"
+  ceph_servers = search(:node, "roles:ceph-osd#{ceph_env_filter}") || []
+  if ceph_servers.length > 0
+    include_recipe "ceph::keyring"
+  else
+    message = "Ceph was not deployed with Crowbar yet!"
+    Chef::Log.fatal(message)
+    raise message
+  end
+end
+
+if has_external
+  # Ensure ceph is available here
+  if node[:platform] == "suse"
+    # install package in compile phase because we will run "ceph -s"
+    package "ceph-common" do
+      action :nothing
+    end.run_action(:install)
+  end
+end
+
+package "qemu-block-rbd" do
+  action :install
+  only_if { node[:platform_version].to_f >= 12.0 }
+end
+
+# Second loop to do our setup
+cinder_controller[0][:cinder][:volumes].each_with_index do |volume, volid|
+  next unless volume[:backend_driver] == "rbd"
+
+  if volume[:rbd][:use_crowbar]
+    ceph_conf = "/etc/ceph/ceph.conf"
+    admin_keyring = "/etc/ceph/ceph.client.admin.keyring"
+  else
+    ceph_conf = volume[:rbd][:config_file]
+    admin_keyring = volume[:rbd][:admin_keyring]
+
+    if ceph_conf.empty? || !File.exists?(ceph_conf)
+      Chef::Log.info("Ceph configuration file is missing; skipping the ceph setup for backend #{volume[:backend_name]}")
+      next
+    end
+
+    if !admin_keyring.empty? && File.exists?(admin_keyring)
+      Chef::Log.info("Using external ceph cluster for nova #{volume[:backend_name]} backend, with automatic setup.")
+    else
+      Chef::Log.info("Using external ceph cluster for nova #{volume[:backend_name]} backend, with no automatic setup.")
+      next
+    end
+
+    cmd = ["ceph", "-k", admin_keyring, "-c", ceph_conf, "-s"]
+    check_ceph = Mixlib::ShellOut.new(cmd)
+
+    unless check_ceph.run_command.stdout.match("(HEALTH_OK|HEALTH_WARN)")
+      Chef::Log.info("Ceph cluster is not healthy; skipping the ceph setup for backend #{volume[:backend_name]}")
+      next
+    end
+  end
+
+  rbd_user = volume[:rbd][:user]
+  rbd_uuid = volume[:rbd][:secret_uuid]
+
+  secret_file_path = "/etc/ceph/ceph-secret-#{rbd_uuid}.xml"
+  
+  file secret_file_path do
+    owner "root"
+    group "root"
+    mode "0640"
+    content "<secret ephemeral='no' private='no'> <uuid>#{rbd_uuid}</uuid><usage type='ceph'> <name>client.#{rbd_user} secret</name> </usage> </secret>"
+  end #file secret_file_path
+
+  ruby_block "save nova key as libvirt secret" do
+    block do
+      # Check if libvirt was installed
+      virsh_hostname = Mixlib::ShellOut.new("virsh hostname")
+      virsh_hostname.run_command
+      virsh_hostname.error!
```

Comment:
That's different from what we had before. In the past, we were using this call to check that libvirt is started, and if it was not the case, we were not doing anything. Now, we are failing in that case, which is pretty bad as I believe we will hit this on initial setup.